File: sort-mapper-example-java/src/main/java/kh/farrukh/sortmapperexamplejava/teacher/TeacherApi.java
package kh.farrukh.sortmapperexamplejava.teacher;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import kh.farrukh.sortmapperexamplejava.teacher.model.TeacherCreateRequestDTO;
import kh.farrukh.sortmapperexamplejava.teacher.model.TeacherDetailsResponseDTO;
import kh.farrukh.sortmapperexamplejava.teacher.model.TeacherResponseDTO;
import kh.farrukh.sortmapperexamplejava.teacher.model.TeacherUpdateRequestDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;

@Tag(name = "Teacher API")
@RequestMapping("/api/v1/teacher")
public interface TeacherApi {

  @Operation(summary = "Get all teachers")
  @GetMapping
  ResponseEntity<Page<TeacherResponseDTO>> getTeachers(Pageable pageable);

  @Operation(summary = "Get teacher by id")
  @GetMapping("{id}")
  ResponseEntity<TeacherDetailsResponseDTO> getTeacher(@PathVariable Long id);

  @Operation(summary = "Create teacher")
  @PostMapping
  ResponseEntity<Void> createTeacher(@Valid @RequestBody TeacherCreateRequestDTO requestDTO);

  @Operation(summary = "Update teacher")
  @PutMapping("{id}")
  ResponseEntity<Void> updateTeacher(
      @PathVariable Long id,
      @Valid @RequestBody TeacherUpdateRequestDTO requestDTO
  );

  @Operation(summary = "Delete teacher")
  @DeleteMapping("{id}")
  ResponseEntity<Void> deleteTeacher(@PathVariable Long id);

}

File: sort-mapper-example/src/main/kotlin/kh/farrukh/sortmapperexample/teacher/TeacherApi.kt
package kh.farrukh.sortmapperexample.teacher

import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.validation.Valid
import kh.farrukh.sortmapperexample.teacher.model.TeacherCreateRequestDTO
import kh.farrukh.sortmapperexample.teacher.model.TeacherDetailsResponseDTO
import kh.farrukh.sortmapperexample.teacher.model.TeacherResponseDTO
import kh.farrukh.sortmapperexample.teacher.model.TeacherUpdateRequestDTO
import org.springframework.data.domain.Pageable
import org.springframework.data.web.PagedModel
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@Tag(name = "Teacher API")
@RequestMapping("/api/v1/teachers")
interface TeacherApi {

    @Operation(summary = "Get all teachers")
    @GetMapping
    fun getTeachers(pageable: Pageable): ResponseEntity<PagedModel<TeacherResponseDTO>>

    @Operation(summary = "Create a new teacher")
    @PostMapping
    fun createTeacher(@Valid @RequestBody requestDTO: TeacherCreateRequestDTO): ResponseEntity<Unit>

    @Operation(summary = "Get teacher by ID")
    @GetMapping("{id}")
    fun getTeacherById(@PathVariable id: Long): ResponseEntity<TeacherDetailsResponseDTO>

    @Operation(summary = "Update teacher by ID")
    @PutMapping("{id}")
    fun updateTeacher(
        @PathVariable id: Long,
        @Valid @RequestBody requestDTO: TeacherUpdateRequestDTO
    ): ResponseEntity<Unit>

    @Operation(summary = "Delete teacher by ID")
    @DeleteMapping("{id}")
    fun deleteTeacher(@PathVariable id: Long): ResponseEntity<Unit>

}